### PR TITLE
Media management

### LINKED ISSUE #118: Image (JPEG/PNG) optimization via e.g. tinyjpg.com
As we store pictures and deliver them throw the api, we should consider to optimize them in size for better network/access performance. We could do this with e.g. tinyjpg.com or other services like that. We should anyway store the original upload from the user but should process this and deliver the optimized version to the clients.

### PR DESCRIPTION
**Pull request by ulliholtgrave**
### Short description
This PR adds a media library to the CMS. For this the library [preact](https://preactjs.com/guide/v10/getting-started) is used, which can render a conditional HTML field into all possible documents. The media library currently has all the core features implemented (upload files, gallery view, change name/alt_text, editor integration), but still needs some love in the UI/UX field.

As we needed to implement a own content validation for setting the correct the media properties, this PR also includes a fix for #604 .

_By accident we fixed #118 as well, as our current implementation reduces images to 300x300px and adds a Link for the user to access the original physical resource by clicking on it._ (TBD)

### Proposed changes
- Add preact to our develop-dependencies
- Remove old media implementations
- Add AJAX-Endpoints for document manipulation
- Adds a region-based check if images can be manipulated (#60)
- Uses Pillow for creating thumbnails and optimization
- Add TinyMCE plugin for media integration
- Add directories for better management
- Add Breadcrumbs for Navigation
- Add Media-Library to the IconWidget for all content types.
- Update TinyMCE settings according to the results of the UI/UX study.

### Resolved issues

Fixes: #48
Fixes: #60 
Fixes: #118 
Fixes: #604 
Fixes: #811
Fixes: #843

### CODE REVIEW

File: src/cms/urls.py [hunk 1]
```diff
@@ -384,6 +385,66 @@
...
+                ),
+                url(
+                    r"^media_edit$",
+                    media.edit_media_element_ajax,
+                    name="mediacenter_admin_edit_url",
+                ),
+                url(
+                    r"^media_edit/create_directory$",
+                    media.create_directory_ajax,
+                    name="mediacenter_admin_create_directory",
```
Comment (timoludwig) [inline src/cms/urls.py:431]:
```suggestion
                    name="mediacenter_create_directory",
```

File: src/cms/views/media/media_actions.py [hunk 2]
```diff
@@ -1,38 +1,259 @@
...
+
+    if "upload" in request.FILES:
+        # TODO: Make sure only allowed files are uploaded by checking mime type here:
+        document = Document()
+        document.region = region
+        document.parent_directory = directory
+        document.name = request.FILES["upload"].name
+        attach_file(document, request.FILES["upload"])
+        document.save()
+        return JsonResponse({"success": True}, status=200)
```
Comment (timoludwig) [inline src/cms/views/media/media_actions.py:57]:
```suggestion
        return JsonResponse({"success": True})
```
The status 200 is the default one.

File: src/cms/views/media/media_actions.py [hunk 3]
```diff
@@ -1,38 +1,259 @@
...
+
+    return JsonResponse({"success": False, "error": _("No file was uploaded")})
```
Comment (timoludwig) [inline src/cms/views/media/media_actions.py:59]:
```suggestion
    return JsonResponse({"success": False, "error": _("No file was uploaded")}, status=400)
```

File: src/cms/views/media/media_list_view.py [hunk 4]
```diff
@@ -1,42 +1,67 @@
...
-    #: The template to render (see :class:`~django.views.generic.base.TemplateResponseMixin`)
     template_name = "media/media_list.html"
     #: The context dict passed to the template (see :class:`~django.views.generic.base.ContextMixin`)
     base_context = {"current_menu_item": "media"}
 
     def get(self, request, *args, **kwargs):
         """
         Render media list
+        """
+        context = self.get_context_data(**kwargs)
```
Comment (timoludwig) [inline src/cms/views/media/media_list_view.py:32]:
```suggestion
    extra_context = {"current_menu_item": "media"}
```
No reason to over-complicate things.

File: src/cms/views/media/media_list_view.py [hunk 5]
```diff
@@ -1,42 +1,67 @@
...
-        :param request: The current request
-        :type request: ~django.http.HttpResponse
+        return render(
+            request,
+            self.template_name,
+            {
+                **self.base_context,
+                **context,
+            },
+        )
```
Comment (timoludwig) [inline src/cms/views/media/media_list_view.py:40]:
```suggestion
```

File: src/cms/views/media/media_list_view.py [hunk 6]
```diff
@@ -1,42 +1,67 @@
...
-        :return: The rendered template response
-        :rtype: ~django.template.response.TemplateResponse
+    template_name = "media/media_list_admin.html"
+    #: The context dict passed to the template (see :class:`~django.views.generic.base.ContextMixin`)
+    base_context = {"current_menu_item": "media"}
+
+    def get(self, request, *args, **kwargs):
+        """
+        Render media list
         """
```
Comment (timoludwig) [inline src/cms/views/media/media_list_view.py:57]:
```suggestion
    extra_context = {"current_menu_item": "media"}
```

File: src/cms/views/media/media_list_view.py [hunk 7]
```diff
@@ -1,42 +1,67 @@
...
-        documents = Document.objects.all()
+        context = self.get_context_data(**kwargs)
 
         return render(
```
Comment (timoludwig) [inline src/cms/views/media/media_list_view.py:60]:
```suggestion
```

File: src/cms/views/media/media_list_view.py [hunk 8]
```diff
@@ -1,42 +1,67 @@
...
-            request, self.template_name, {**self.base_context, "documents": documents}
+            request,
+            self.template_name,
+            {
+                **self.base_context,
+                **context,
+            },
         )
```
Comment (timoludwig) [inline src/cms/views/media/media_list_view.py:67]:
```suggestion
```

File: src/cms/views/media/media_actions.py [hunk 9]
```diff
@@ -1,38 +1,259 @@
...
+
+    json_data = json.loads(request.body.decode("utf-8"))
+    media_element_id = json_data["id"]
+    media_element = get_object_or_404(Document, id=media_element_id)
+    if media_element.region != region:
+        raise PermissionDenied
+
+    media_element.delete()
+
+    return JsonResponse({"success": True}, status=200)
```
Comment (timoludwig) [inline src/cms/views/media/media_actions.py:95]:
```suggestion
    return JsonResponse({"success": True})
```

File: src/cms/views/media/media_actions.py [hunk 10]
```diff
@@ -1,38 +1,259 @@
...
+    directory = Directory.objects.get(id=directory_id)
+    if directory.region is not None and directory.region != region:
+        raise PermissionDenied
+
+    directory_path = []
+    while directory is not None:
+        directory_path.insert(0, directory.serialize())
+        directory = directory.parent
+
+    return JsonResponse({"success": True, "data": directory_path}, status=200)
```
Comment (timoludwig) [inline src/cms/views/media/media_actions.py:129]:
```suggestion
    return JsonResponse({"success": True, "data": directory_path})
```

File: src/cms/views/media/media_actions.py [hunk 11]
```diff
@@ -1,38 +1,259 @@
...
+    media_element_id = json_data["id"]
+    media_element = get_object_or_404(Document, id=media_element_id)
+    if media_element.region != region:
+        raise PermissionDenied
+
+    media_element.name = json_data["name"]
+    media_element.alt_text = json_data["alt_text"]
+    media_element.save()
+
+    return JsonResponse({"success": True}, status=200)
```
Comment (timoludwig) [inline src/cms/views/media/media_actions.py:210]:
```suggestion
    return JsonResponse({"success": True})
```

File: src/cms/models/media/document.py [hunk 12]
```diff
@@ -35,7 +69,6 @@ def delete(self, using=None, keep_parents=False):
         :return: The number of objects deleted and a dictionary with the number of deletions per object type.
         :rtype: tuple ( int, dict )
         """
-        self.document.delete()
```
[the remark this answers is not in the document]
Comment (ulliholtgrave) [reply src/cms/models/media/document.py:38]: You are right. Removed it in the last commits.

File: src/cms/views/media/media_actions.py [hunk 13]
```diff
@@ -1,38 +1,274 @@
...
+    if int(json_data["parentDirectory"]) != 0:
+        parent_directory = get_object_or_404(Directory, id=json_data["parentDirectory"])
+
+        if parent_directory.region != region:
+            raise PermissionDenied
+    else:
+        json_data["parentDirectory"] = None
+
+    if Directory.objects.filter(
+        parent=json_data["parentDirectory"], name=json_data["directoryName"]
```
Comment (timoludwig) [inline src/cms/views/media/media_actions.py:264]:
```suggestion
    queryset = Directory.objects
    if region:
        # If directory is created for a region, only limit choices to the global library and the regional one
        queryset = queryset.filter(Q(region=region) | Q(region__isnull=True))
    if queryset.filter(
        parent=parent_directory,
        name=json_data["directoryName"],
```
I think it's enough if the directory name is unique across the global library and the region's library...
If two separate regions want to create directories with the same name, that should be fine I think...

File: src/cms/views/media/media_actions.py [hunk 14]
```diff
@@ -1,38 +1,274 @@
...
+
+    json_data = json.loads(request.body.decode("utf-8"))
+    parent_directory = None
+    if int(json_data["parentDirectory"]) != 0:
+        parent_directory = get_object_or_404(Directory, id=json_data["parentDirectory"])
+
+        if parent_directory.region != region:
+            raise PermissionDenied
+    else:
+        json_data["parentDirectory"] = None
```
Comment (timoludwig) [inline src/cms/views/media/media_actions.py:261]:
```suggestion
        parent_directory = get_object_or_404(
            Directory.objects.filter(region=region), id=json_data["parentDirectory"]
        )
```

File: src/cms/views/media/media_actions.py [hunk 15]
```diff
@@ -1,38 +1,274 @@
...
+        return JsonResponse({"error": _("Directory already exists")}, status=400)
 
-    if request.method == "POST":
-        document = Document.objects.get(id=document_id)
-        document.delete()
+    new_directory = Directory()
+    new_directory.name = json_data["directoryName"]
+    new_directory.parent = parent_directory
+    new_directory.region = region
+    new_directory.save()
```
Comment (timoludwig) [inline src/cms/views/media/media_actions.py:272]:
```suggestion
    Directory.objects.create(
        name=json_data["directoryName"], parent=parent_directory, region=region
    )
```

File: src/cms/models/media/directory.py [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+        return f"<Directory (id: {self.id}{region}, name: {self.name})>"
+
+    class Meta:
+        """
+        This class contains additional meta configuration of the model class, see the
+        `official Django docs <https://docs.djangoproject.com/en/2.2/ref/models/options/>`_ for more information.
+        :param default_permissions: The default permissions for this model
+        :type default_permissions: tuple
+        """
+
```
Comment (timoludwig) [inline src/cms/models/media/directory.py:68]:
```suggestion
        #: The verbose name of the model
        verbose_name = _("directory")
        #: The plural verbose name of the model
        verbose_name_plural = _("directories")
        #: The default permissions for this model
```

File: src/cms/models/media/document.py [hunk 17]
```diff
@@ -65,3 +98,7 @@ class Meta:
         verbose_name_plural = _("documents")
         #: The default permissions for this model
         default_permissions = ()
+        permissions = (
```
Comment (timoludwig) [inline src/cms/models/media/document.py:101]:
```suggestion
        #:  The custom permissions for this model
        permissions = (
```

File: src/cms/models/media/directory.py [hunk 18]
```diff
@@ -0,0 +1,69 @@
+from django.db import models
```
Comment (timoludwig) [inline src/cms/models/media/directory.py:1]:
```suggestion
from django.db import models
from django.utils.translation import ugettext_lazy as _

```

File: src/cms/models/media/directory.py [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+from cms.models.regions.region import Region
```
Comment (timoludwig) [inline src/cms/models/media/directory.py:2]:
```suggestion
from ..regions.region import Region
```

File: src/cms/models/media/directory.py [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+
+
+class Directory(models.Model):
+    """
+    Model representing a directory containing documents. This is only a virtual directory and does not necessarily
+    exist on the actual file system. Each directory is tied to a region.
+    """
+
+    name = models.CharField(max_length=255, blank=False)
```
Comment (timoludwig) [inline src/cms/models/media/directory.py:11]:
```suggestion
    name = models.CharField(max_length=255, blank=False, verbose_name=_("name"))
```

File: src/cms/models/media/directory.py [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+    region = models.ForeignKey(
+        Region,
+        related_name="media_directories",
+        on_delete=models.CASCADE,
+        null=True,
+        blank=True,
```
Comment (timoludwig) [inline src/cms/models/media/directory.py:17]:
```suggestion
        blank=True,
        verbose_name=_("region"),
```

File: src/cms/models/media/directory.py [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+    )
+    parent = models.ForeignKey(
+        "self",
+        related_name="subdirectories",
+        on_delete=models.PROTECT,
+        blank=True,
+        null=True,
```
Comment (timoludwig) [inline src/cms/models/media/directory.py:24]:
```suggestion
        null=True,
        verbose_name=_("parent directory"),
```

File: src/cms/models/media/document.py [hunk 23]
```diff
@@ -1,52 +1,84 @@
...
     """
-    The Document model is used to store meta-data about files which are uploaded to the CMS.
+    The Document model is used to store basic information about files which are uploaded to the CMS. This is only a
+    virtual document and does not necessarily exist on the actual file system. Each document is tied to a region via its
+    directory.
     """
 
-    description = models.CharField(
-        max_length=255, blank=True, verbose_name=_("description")
+    physical_path = models.CharField(max_length=255, blank=False)
```
Comment (timoludwig) [inline src/cms/models/media/document.py:20]:
```suggestion
    physical_path = models.CharField(
        max_length=255,
        blank=False,
        verbose_name=_("physical file path"),
    )
```

File: src/cms/models/media/document.py [hunk 24]
```diff
@@ -1,52 +1,84 @@
...
+    type = models.CharField(max_length=255, blank=False)
+
+    name = models.CharField(max_length=255, blank=False, verbose_name=_("name"))
+    parent_directory = models.ForeignKey(
+        Directory, related_name="documents", on_delete=models.PROTECT, null=True
```
Comment (timoludwig) [inline src/cms/models/media/document.py:25]:
```suggestion
        Directory,
        related_name="documents",
        on_delete=models.PROTECT,
        null=True,
        verbose_name=_("parent directory"),
```

File: src/cms/models/media/document.py [hunk 25]
```diff
@@ -1,52 +1,84 @@
...
-    document = models.FileField(
-        upload_to="",
-        verbose_name=_("document"),
-        help_text=_("The actual document file"),
+    region = models.ForeignKey(
+        Region,
+        related_name="documents",
+        on_delete=models.CASCADE,
+        null=True,
+        blank=True,
```
Comment (timoludwig) [inline src/cms/models/media/document.py:32]:
```suggestion
        blank=True,
        verbose_name=_("region"),
```

File: src/cms/models/media/document.py [hunk 26]
```diff
@@ -1,52 +1,84 @@
...
     )
+    uploaded_at = models.DateTimeField(auto_now_add=True)
```
Comment (timoludwig) [inline src/cms/models/media/document.py:34]:
```suggestion
```
There's already an `uploaded_at` field a few lines below.

**Review by timoludwig — COMMENTED** [review]
I tried to optimize a few things, I hope this doesn't break everything :smile: :innocent:

File: src/cms/templates/_base.html [hunk 27]
```diff
@@ -127,22 +127,10 @@
...
-                    <a href="{% url 'new_upload_file' document_id='0' region_slug=region.slug %}"
-                       class="{% if current_menu_item == 'media_upload' %} active {% endif %}">
-                        {% trans 'Upload new media file' %}
-                    </a>
-                </div>
-            </div>
+            <a href="{% url 'media' region_slug=region.slug %}" class="{% if current_menu_item == 'media' %} active {% endif %}">
+                <i data-feather='file'></i>
+                {% trans 'Media Library' %}
+            </a>
```
[the remark this answers is not in the document]
Comment (ulliholtgrave) [reply src/cms/templates/_base.html:133]: Added it again in local commit.

**Review by timoludwig — APPROVED** [review]
Sorry that it took me so long to implement the improvements I had in mind for this PR!
See the commit description for a detailed list of what I changed.

Thanks again  @ulliholtgrave and especially @jnugh for your great work and I'm really looking forward to merge this PR :tada: :100:

**Pull request merged**